### ISSUE TITLE
✅ 📊 🐞 Bug Report: Mega Menu super slow low loading? - [ED-10424]

### ISSUE THREAD

**Issue by nesoor** [elementor/elementor#21956]
### Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [ ] The issue still exists against the latest stable version of Elementor.

### Description

The mega menu performs well initially, but as more dynamic internal URLs are added, it becomes significantly slower. This issue noticeably affects the loading time and disrupts the menu editing workflow. The loading process takes around 3 minutes, and adding an additional link also takes a considerable amount of time. It is important to note that this problem is specific to Elementor's Mega Menu, as the rest of the editing process remains fast and smooth without any complications.

### Steps to reproduce

After adding approximately 276 links to the mega menu, its performance has been significantly impacted.

### Isolating the problem

- [ ] This bug happens with only Elementor plugin active (and Elementor Pro).
- [X] This bug happens with a Blank WordPress theme active ([Hello theme](https://wordpress.org/themes/hello-elementor/)).
- [X] I can reproduce this bug consistently following the steps above.

### System Info

<details>
<summary>Click to reveal</summary>

````txt



== Server Environment ==
	Operating System: Linux
	Software: Apache/2
	MySQL version: MariaDB Server v10.4.27
	PHP Version: 8.2.2
	PHP Memory Limit: 2048M
	PHP Max Input Vars: 1000
	PHP Max Post Size: 64M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: All right
	Elementor Library: Connected

== WordPress Environment ==
	Version: 6.2
	Site URL: https://www.domeinnaamhier.nl
	Home URL: https://www.domeinnaamhier.nl
	WP Multisite: No
	Max Upload Size: 32 MB
	Memory limit: 40M
	Max Memory limit: 2048M
	Permalink Structure: /%category%/%postname%/
	Language: nl
	Timezone: Europe/Amsterdam
	Debug Mode: Inactive

== Theme ==
	Name: Hello Elementor Child
	Version: 1.0.0
	Author: Elementor Team
	Child Theme: Yes
	Parent Theme Name: Hello Elementor
	Parent Theme Version: 2.7.1
	Parent Theme Author: Elementor Team

== User ==
	Role: administrator
	WP Profile lang: nl_NL
	User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36

== Active Plugins ==
	Advanced Access Manager
		Version: 6.9.8
		Author: Vasyl Martyniuk <vasyl@vasyltech.com>

	All In One WP Security
		Version: 5.1.7
		Author: All In One WP Security & Firewall Team

	Complianz | GDPR/CCPA Cookie Consent
		Version: 6.4.3
		Author: Really Simple Plugins

	Elementor
		Version: 3.12.1
		Author: Elementor.com

	Elementor Pro
		Version: 3.12.1
		Author: Elementor.com

	EWWW Image Optimizer
		Version: 6.9.3
		Author: Exactly WWW

	Post Types Order
		Version: 2.0.2
		Author: Nsp Code

	PowerPack Pro for Elementor
		Version: 2.9.17
		Author: Team IdeaBox - PowerPack Elements

	Rank Math SEO
		Version: 1.0.111
		Author: Rank Math

	Rank Math SEO PRO
		Version: 3.0.33
		Author: Rank Math

	SWUP Webdesign
		Version: 2.0
		Author: SWUP Webdesign

	WP Super Cache
		Version: 1.9.4
		Author: Automattic


== Must-Use Plugins ==
	Elementor Safe Mode
		Version: 1.0.0
		Author: Elementor.com


== Eigenschappen ==
	Custom Fonts: 0
	Custom Icons: 0

== Integrations ==
	
	recaptcha_v3: Active


== Elementor experimenten ==
	Geoptimaliseerde DOM-uitvoer: Actief
	Verbeterd laden van assets: Actief
	Verbeterd laden van CSS: Actief
	Pictogrammen inline lettertype: Actief
	Toegang verbeteringen: Actief
	Extra aangepaste onderbrekingspunten: Actief
	admin_menu_rearrangement: Standaard inactief
	Flexbox container: Actief
	Upgrade Swiper bibliotheek: Actief
	Default to New Theme Builder: Actief
	Hello Theme header &amp; footer: Actief
	Editor top balk: Actief
	Landingspagina&#039;s: Actief
	Geneste elementen: Actief
	Lazy-load achtergrondafbeeldingen: Actief
	Page Transitions: Actief
	Notes: Actief
	Loop: Actief
	Form Submissions: Actief
	Scroll Snap: Actief
	Menu: Actief


== Log ==
	
Log: showing 20 of 362023-03-29 19:00:46 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.11.5',
  'to' => '3.12.0',
)]
2023-03-29 19:00:47 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.11.7',
  'to' => '3.12.0',
)]
2023-03-29 19:00:48 [info] elementor-pro::elementor_pro_updater Started 
2023-03-29 19:00:48 [info] Elementor Pro/Upgrades - _on_each_version Start  
2023-03-29 19:00:48 [info] Elementor Pro/Upgrades - _on_each_version Finished 
2023-03-29 19:00:48 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.11.7',
  'to' => '3.12.0',
)]
2023-03-29 19:00:48 [info] elementor::elementor_updater Started 
2023-03-29 19:00:48 [info] Elementor/Upgrades - _on_each_version Start  
2023-03-29 19:00:48 [info] Elementor/Upgrades - _on_each_version Finished 
2023-03-29 19:00:48 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.11.5',
  'to' => '3.12.0',
)]
2023-04-03 13:52:46 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.12.0',
  'to' => '3.12.1',
)]
2023-04-03 13:52:47 [info] elementor-pro::elementor_pro_updater Started 
2023-04-03 13:52:47 [info] Elementor Pro/Upgrades - _on_each_version Start  
2023-04-03 13:52:47 [info] Elementor Pro/Upgrades - _on_each_version Finished 
2023-04-03 13:52:47 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.12.0',
  'to' => '3.12.1',
)]
2023-04-03 13:52:47 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.12.0',
  'to' => '3.12.1',
)]
2023-04-03 13:52:47 [info] elementor::elementor_updater Started 
2023-04-03 13:52:47 [info] Elementor/Upgrades - _on_each_version Start  
2023-04-03 13:52:47 [info] Elementor/Upgrades - _on_each_version Finished 
2023-04-03 13:52:47 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.12.0',
  'to' => '3.12.1',
)]

PHP: showing 2 of 2PHP: 2023-03-15 14:23:31 [warning X 100][../wp-content/plugins/elementor-pro/modules/query-control/module.php::749] Attempt to read property &quot;name&quot; on null [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2023-04-05 10:25:11 [warning X 1][../wp-content/plugins/elementor/core/common/modules/ajax/module.php::175] Undefined array key &quot;data&quot; [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]

JS: showing 4 of 4JS: 2023-04-02 11:26:12 [error X 11][https://www.domeinnaamhiernl/wp-content/plugins/elementor/assets/js/editor.min.js?ver=3.12.0:3:388090] Cannot read properties of undefined (reading &#039;controls&#039;) 
JS: 2023-04-03 12:34:59 [error X 1][https://www.domeinnaamhier.nl/wp-content/plugins/elementor/assets/js/editor.min.js?ver=3.12.1:3:785238] Cannot read properties of undefined (reading &#039;toLowerCase&#039;) 
JS: 2023-04-04 18:45:53 [error X 14][https://www.domeinnaamhier.nl/wp-content/plugins/elementor/assets/js/df16c9d56464dd4f0800.bundle.min.js:2:7826] Cannot read properties of null (reading &#039;dataset&#039;) 
JS: 2023-04-05 10:19:42 [error X 1][https://www.domeinnaamhier.nl/wp-content/plugins/elementor/assets/js/editor.min.js?ver=3.12.1:3:571297] Element type not found: &#039;ekit-nav-menu&#039; 



== Elementor - Compatibility Tag ==
	
	Elementor Pro: Compatibiliteit niet gespecificeerd
	PowerPack Pro for Elementor: Compatibiliteit niet gespecificeerd

== Elementor Pro - Compatibility Tag ==
	



````

</details>

**Comment by oliveandlake** [comment]
I have exactly the same issue with the Mega Menu. However, it's not consistent. Sometimes the page will load quickly, sometimes it won't. It can be up to a couple of minutes before anything happens. There's nothing in dev tools 'network' either, it's just waiting. It will also affect the editor loading too. Eventually it does load, but it can take quite a time. The issue is that it's sporadic. Sometimes it will load quickly and other times it doesn't. 

I have a nav menu with 5 items. One is a mega menu, the rest are normal. In the one mega menu, I have a tab with about 8 tabs. Screenshot provided. 

![Screenshot_4](https://user-images.githubusercontent.com/56640831/230727519-ec519032-73d7-4ae0-99a4-244c7d4e6382.png)

Here is a video of it in action:

https://www.loom.com/share/e5ff81fc6a6446b9bc421c8abda355f4

**Comment by nicholaszein** [comment]
Hello there! 

Thank you for submitting your issue! 🙏 

We are aware of this issue and our team is already working on a fix. We can’t say exactly when it'll be sorted out, but we kindly ask for your patience. Thanks for understanding! 😉

Best regards

**Comment by aovivo2000** [comment]
I fear I have the same issue... 

I am working on a staging site on a managed server with usually 0.6 s loading time (having filter pluiginms, Woocommerce etc installed) , 
I am using Elementor now and its Mega menu the fully loaded site sometimes takes up to 8 seconds! As the author above described that is sometimes very different, some times 7 secondfs, sometimes 5 seconds, but its never unter 3 seconds... 

The preformance issues still persist, even if I use an empty page, with only Mega Menu and Footer running
Hope that we can to the bottom of this... since this is a bit frustrating
thanks!
P.-

**Comment by TOTLD** [comment]
+1 this issue seems to continue to exist. 

Suggested steps about caching disable, Improved CSS Loading feature disable, regenerating files and purging all caches does make the editing process slightly smoother momentarily but after one Publish iteration is is slowing right down again.

Front-end performance is average but this could be a reflection of having to disable a lot of optimisation tools just for the Mega Menu to work at all. Far from ideal as this affects much more than just the menu.

6 months since this report here, and this feature is crucial to complex site menu designs. Hopefully it is genuinely being worked on.

Cheers,
Tom

**Comment by barttermorshuizen** [comment]
+1.  Seems to be tied to the autocomplete mechanism in the link field. The entire interface becomes sluggish in editorial. Important to get this fixed / it’s a great widget if it works!

**Comment by feriamsidera** [comment]
Still an issue even on a local setup running version 3.18 - autocomplete runs slow as molasses in the winter time. Widget runs fine once edited. Loading the editor with any page running the mega menu takes much longer than normal as well. Please fix.

**Comment by midihead7** [comment]
+1. Not only is the editor still very slow when editing the mega menu, **it is now impossible for me to add a new parent link to the menu and save the changes**. Every time I try, the editor gets even slower, and saving the changes is no longer possible. The editor just freezes completely upon saving and I have to just close the browser tab to get out of the editor. This happens with only Elementor, Elementor Pro, and the Hello theme active. **Please fix this ASAP!**

**Comment by barttermorshuizen** [comment]
Hi, any news on progress here? It's really important to get this fixed!

**Comment by khdigital94** [comment]
Seems like the devs don't care about this issue anymore. Last response in early September. Great

**Comment by brunozanoelo** [comment]
Any update? This feature is impossible to work with, this bug must be solved guys. I already spent hours trying to solve or work with mega menu in elementor (as the marketing says it is easy)

**Comment by brunozanoelo** [comment]
This is a HUGE impact in our Elementor project. We are only asking for an UPDATE w/ release dates!

**Comment by webdevs-pro** [comment]
Nested elements is very slow in editor

**Comment by TOTLD** [comment]
Hoping we may see a mega-menu improvement in a coming update?

I hadn't used this feature since 3.18.1, but have just tried again on 3.18.3 and Pro 3.18.2 and note there is no noticeable change.

Working with the mega-menu is still quite unreliable and causes hang within the Editor regularly. Chrome appears to suffer less than FireFox Dev Edition, where it is totally unusable.

**Comment by brunozanoelo** [comment]
I´ve updated Elementor and Wordpress and NOW is working fine for me, sample:
https://systempro.com.br/

**Comment by midihead7** [comment]
The editor in the latest versions (free 3.18.3 and Pro 3.18.2) still freezes the editor when trying to add a new link to the menu. The editor refuses to save after adding a new URL to the menu. I opened Case# TSJG9TJF14AK with Elementor Support on December 7, 2023, for this issue. They escalated it to Tier 3 support and then gave me the generic answer **"I have received a response from Tier 3, and we have created an internal report regarding this issue. Hopefully, this will be fixed in a future version of Elementor. Though we don't have an ETA for the same, please keep an eye on the changelog to see when this issue gets resolved. Ref: https://elementor.com/pro/changelog/".** They immediately closed the ticket after sending their reply. Support saying that they _hope_ to have it fixed isn't the same thing as acknowledging that it is a confirmed bug. This issue was reported here back in April of 2023. This has been an issue for far too long and needs to be resolved quickly.

**Comment by momo-fr** [comment]
Same issue for me, I'm trying to create a mega menu and after activating the function on 2 menus it became extremely slow. Editing sometimes gets stuck. I have to wait 1 to minutes to be able to modify something. It's unusable!!!

**Comment by magicmazuk** [comment]
Same issue here. As soon as I add a forth menu item the page just grinds to a halt. Something I did notice and not sure if it is related. When I add menu items, by default it goes: "Item #1", "item #2", "item #3", But when I go to add another menu (when it dies), it adds a new menu item called "item #3" again. Duplicating the name. Maybe the duplicated object is causing the memory leak? Just a thought.

**Comment by Dudeney95** [comment]
+1 Same issue here

**Comment by nicolasstrands** [comment]
+1 Encountering the same issue here. Initially implemented that on a client's website but we ended up having too many issues with the widget so we had to take it out entirely.

**Comment by momo-fr** [comment]
It's time to find a solution, it's almost unusable!!

**Comment by barttermorshuizen** [comment]
any of you find a good alternative? 

One that doesn't work is the jetmenu - it has blocking issues with WPML and doesn't work properly with opening on hover.

really unacceptable that this issue is open for 10 months!!!!!!!!

**Comment by magicmazuk** [comment]
> any of you find a good alternative?
> 
> One that doesn't work is the jetmenu - it has blocking issues with WPML and doesn't work properly with opening on hover.
> 
> really unacceptable that this issue is open for 10 months!!!!!!!!

We ended up using https://wordpress.org/plugins/elementskit-lite/. Its a bit fiddly but got there. See it in action here if you want: https://www.hcicontracts.com/

**Comment by nicholaszein** [comment]
@momo-fr a solution is on the works already. We will update this thread when we have it ready!

**Comment by pg-los** [comment]
@nicholaszein nice to know you are investigating this issue. I too am experiencing the same problem and the widget has become uneditable.
Let us know as soon as you find a solution, thank you!

**Comment by Erudition** [comment]
For me, the menu widget seems to be bringing the page to a crawl, and after adding the fourth menu item, the page can't be saved/published (it just spins forever)

**Comment by barttermorshuizen** [comment]
This bug, that block all practical use of the widget, is 2 weeks away from its first birthday!

Come on Elementor! This takes way too long to fix!

**Comment by pg-los** [comment]
A workaround seems to be using safari on Mac when editing the mega menu. Much faster. But obviously not a solution. I hope this helps somebody. CheersPier Giorgio ProvenzanoHead of DigitalThe Lexiconhttps://thelexicon.orgOn 23 Mar 2024, at 01:46, Erudition ***@***.***> wrote:﻿
For me, the menu widget seems to be bringing the page to a crawl, and after adding the fourth menu item, the page can't be saved/published (it just spins forever)

**Comment by barttermorshuizen** [comment]
That’s a great tip! Thanks @pg-los

**Comment by magicmazuk** [comment]
> A workaround seems to be using safari on Mac when editing the mega menu. Much faster. But obviously not a solution. I hope this helps somebody. CheersPier Giorgio ProvenzanoHead of DigitalThe Lexiconhttps://thelexicon.orgOn 23 Mar 2024, at 01:46, Erudition ***@***.***> wrote:﻿ For me, the menu widget seems to be bringing the page to a crawl, and after adding the fourth menu item, the page can't be saved/published (it just spins forever) —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

No difference on safari for me. Still complete lockdown on 4th menu item.

**Comment by eraybektascomtr** [comment]
Although I tried Elementor Pro & every version of Elementor (including beta and developer), none of them could offer a solution.

**Comment by pg-los** [comment]
> Hey guys, over a year since this was flagged and Mega Menu is part of your release promo's but it still does not work as advertised. Anyone who is coming to Elementor to use this feature will be sorely disappointed and the brand's trust is damaged.
> 
> I'm still hitting the Mega Menu's functional limit (e.g. cannot add more dropdowns, Editor stops working properly) with as few as 4 top-level items.
> 
> Is there any updated advice about how to mitigate issues until a working release? Or, if there are technical limitations, can we know about them to make informed decisions about using this "feature" in projects?

Ditto... waiting for a year, but Elementor doesn't seem to care or be able to fix. I guess this is related to the nested elements and mega menu. The issue only bogs down the editor, while online the menu appears to be nice and responsive, although i haven't run any comparative speed tests on load.

It's time Elementor fixed this issue, because this can only get worse if not addressed: the nested flex containers seem to be the way to go as they are slowly ditching support for sections and inner sections. This is a very critical issue, at the core of Elementor, which makes us all frustrated and thinking of getting away from the platform in the near future

**Comment by pg-los** [comment]
> @momo-fr a solution is on the works already. We will update this thread when we have it ready!

any news, please?

**Comment by LMcGhee-zz** [comment]
This is a huge slow down for making progress with my builds. I do LOVE the new menu and how you can customize it anyway you want but opening it up and saving it can take hours. I have websites on both the Elementor hosting and on high powered dedicated servers and both take FOREVER to edit and save. Any word?

**Comment by TravistyTrav** [comment]
Been facing this issue for months now.. thought i was the only one. 

Please prioritize this!

**Comment by mofolo** [comment]
> This is a huge slow down for making progress with my builds. I do LOVE the new menu and how you can customize it anyway you want but opening it up and saving it can take hours. I have websites on both the Elementor hosting and on high powered dedicated servers and both take FOREVER to edit and save. Any word?

It's probably not an issue from the server itself; it looks to be an issue with the Javascript.

**Comment by barttermorshuizen** [comment]
come on guys! this issue is in its second year, blocking its intended use. 

is there anything I can do to raise the priority? Set up a call with someone? Buy drinks for the team? Sending daily messages on this topic?

**Comment by janjcwebtech** [comment]
Running into this issue on the latest version (3.21.X). Seemed ok, but as I was styling and adding content it was gradually slower and slower until changes would not save any more. Tried to disable all plugins, reimport the template, remove elements, import it to another site. No luck. Elementor you made a huge deal of adding this feature, one year later it is unusable!

**Comment by iEscapedVim** [comment]
@nicholaszein any updates?

**Comment by nicholaszein** [comment]
@iEscapedVim,

we are working on a huge update for all nested elements that should improve performance significantly. We expect to finish it soon but not with the release of 3.22.0 yet.

**Comment by AndersJpg** [comment]
+1 im so close to be done with this menu and then find out that it cant handle so many urls and now its just so deadly slow / broken.

**Comment by midihead7** [comment]
@nicholaszein,

If Elementor 3.23.1 and Elementor Pro 3.23.0 are supposed to include the fix for the slow mega menu issues relating to nested elements, my testing on a staging site shows no absolutely no improvement at all. Editing a mega menu is every bit as slow as before.

**Comment by nicholaszein** [comment]
Hello there!

We have great news! 🙌 

### 📢 We're happy to announce that the issue you raised was resolved with the release of Elementor v3.23.0! 🥳 

The time to interaction should improve by 60%-98%, and performance overall should be better. For a full breakdown please read [this article](https://developers.elementor.com/elementor-3-23-developers-update/#:~:text=Performant%20Nested%20Elements%20In%20the%20Editor).

✅ Feel free to check it out and update your plugin to the new version!

> _Check the changelog for more information:_
> [Changelog of Elementor and Elementor Pro](https://elementor.com/help/changelog-of-elementor/)

Cheers 🥂
